src/connection_manager.rs

```diff
@@ -366,4 +411,96 @@ mod test {
+        let run_cm = |stats: Arc<Mutex<Stats>>, o: Receiver<Event>| {
+            spawn(move || {
+                for i in o.iter() {
+                    let mut stats = stats.lock().unwrap();
+                    match i {
+                        Event::NewConnection(other_ep) => {
+                            println!("Connected to --> {:?}", other_ep);
+                            stats.new_connections_count += 1;
+                        },
+                        Event::NewMessage(from_ep, data) => {
+                            println!("New message from {:?} data:{:?}",
+                                     from_ep, decode::<String>(data));
+                            stats.messages_count += 1;
+                            if stats.messages_count == MESSAGE_PER_NODE * (NETWORK_SIZE - 1) {
+                                break;
+                            }
+                        },
+                        Event::LostConnection(other_ep) => {
+                            println!("Lost connection to {:?}", other_ep);
+                            stats.lost_connection_count += 1;
+                        }
+                    }
+                }
+                println!("done");
+            })
+        };
+
+        let mut network = Network { nodes: Vec::new() };
+        let mut stats = Vec::new();
+        let mut runners = Vec::new();
+        for _ in 0..NETWORK_SIZE {
+        let receiver = network.add();
+            let stat = Arc::new(Mutex::new(Stats {new_connections_count: 0, messages_count: 0,
+                 lost_connection_count: 0} ));
+            let stat_copy = stat.clone();
+            let runner = run_cm(stat_copy, receiver);
+            stats.push(stat);
+            runners.push(runner);
+        }
+
+        let mut listening_end_points = Vec::new();
+            for node in network.nodes.iter() {
+            listening_end_points.push(get_endpoint(node));
+        }
+
+        for node in network.nodes.iter() {
+            for end_point in listening_end_points.iter().filter(|&ep| get_endpoint(node).ne(ep)) {
+                let node = node.clone();
+                let ep = end_point.clone();
+                spawn(move || {
+                    let node = node.lock().unwrap();
+                    node.conn_mgr.connect(vec![ep]);
+                });
+            }
+        }
+
+        for node in network.nodes.iter() {
+            for end_point in listening_end_points.iter().filter(|&ep| get_endpoint(node).ne(ep)) {
+                for _ in 0..MESSAGE_PER_NODE {
+                    let node = node.clone();
+                    let ep = end_point.clone();
+                    spawn(move || {
+                        let node = node.lock().unwrap();
+                        let _ = node.conn_mgr.send(ep.clone(), encode(&"message".to_string()));
+                    });
+                }
+            }
+        }
+
+        thread::sleep_ms(100 * NETWORK_SIZE + 10 * MESSAGE_PER_NODE);
```

Comment:
I am not sure the channel give us any benefits. The thread runs until it reaches a known limit. The sleep can be tuned a bit by adding some logic to investigate the Vec<Stats> to see if all messages are received.

The introduction of channel, requires looping on rx.iter()s for transmissions from thread. Then there is the need for putting time out on how long main thread should wait for the threads to join